keps/sig-node/5307-container-restart-policy/README.md

```diff
@@ -0,0 +1,1017 @@
+on a specific exit code.
+- Allow extensibility of an API to support more scenarios in future.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Implement the ​​maxRestartTimes https://github.com/kubernetes/enhancements/issues/3322
+- Support all possible restart policy rules in this KEP - some may be ideas for the future
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+The Pod with two containers - one is the main container and one is the sidecar
+container. The Pod represents one of the pods in a large AI/ML training workload.
+As training jobs are similar in nature, processing of a next job may be done using
+the same Pods. Sidecar container coordinates the jobs. Once a new job needs to be
+started, sidecar containers are killing the main job, and expect that the container
+will be restarted quickly and pick up a new task to execute.
+
+Since those jobs are often declared as a Job, with [pod failure policy](https://kubernetes.io/docs/tasks/job/pod-failure-policy/),
+the error that was the result of a sidecar killing the container should be
+ignored on kubelet and container needs to be restarted quickly.
+
+See https://github.com/kubernetes-sigs/jobset/issues/876 for the detailed description.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+The container level restart policy may restart the container differently from
+the existing assumptions on pod level restart policy. Pod with `restartPolicy=Never`
+may have containers restarted and have the restart count higher than 0. This may
+also affect how Job `podFailurePolicy` interacts with pod failures.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Controllers managing pod failures
```

Comment:
TBH, I don't think this is a risk, because Job controller with podFailurePolicy only looks at failed pods already, when "restartPolicy: Never". It could be a risk if we used "restartPolicy: OnFailure" at the Pod's level. 

Instead of the risk I would suggest expanding the "Notes/Constraints/Caveats" with that comment.